src/Collection.php

```diff
@@ -94,7 +96,13 @@ public static function range($start = 0, $end = null, $step = 1)
     public function getIterator()
     {
         if ($this->inputFactory) {
-            $this->input = call_user_func($this->inputFactory);
+            $input = call_user_func($this->inputFactory);
+
+            if (is_array($input)) {
+                $input = new RecursiveArrayIterator($input);
+            }
+
```

Comment:
Indeed. In addition, commonization of the rule might be more good (by separation to another method).